Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyHistoryCalculations.kt

```diff
@@ -1,39 +1,73 @@
+    private val daysSinceOnsetOfSymptomsVectorDeterminator: DaysSinceOnsetOfSymptomsVectorDeterminator,
     private val keyConverter: KeyConverter
 ) {
 
+    companion object {
+        const val MAX_AGE_IN_DAYS = 14
+        const val TEN_MINUTES_IN_MILLIS = (10 * 60 * 1000).toLong()
+    }
+
     fun transformToKeyHistoryInExternalFormat(
         keys: List<TemporaryExposureKey>,
         symptoms: Symptoms
     ) =
         toExternalFormat(
-            toSortedHistory(limitKeyCount(keys)),
-            transmissionRiskVectorDeterminator.determine(symptoms)
+            removeOldKeys(sortWithRecentKeyFirst(keys)),
+            transmissionRiskVectorDeterminator.determine(symptoms),
+            daysSinceOnsetOfSymptomsVectorDeterminator.determine(symptoms)
         )
 
-    fun <T> limitKeyCount(keys: List<T>): List<T> =
-        keys.take(MAXIMUM_KEYS)
-
-    fun toExternalFormat(
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun removeOldKeys(
         keys: List<TemporaryExposureKey>,
-        transmissionRiskVector: TransmissionRiskVector
+        now: DateTime = Instant().toDateTime(DateTimeZone.UTC)
     ) =
-        keys.mapIndexed { index, key ->
-            // The latest key we receive is from yesterday (i.e. 1 day ago),
-            // thus we need use index+1
-            keyConverter.toExternalFormat(key, transmissionRiskVector.getRiskValue(index + 1))
+        keys.filter { it.ageInDays(now) in 0..MAX_AGE_IN_DAYS }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun toExternalFormat(
+        keys: List<TemporaryExposureKey>,
+        transmissionRiskVector: TransmissionRiskVector,
+        daysSinceOnsetOfSymptomsVector: DaysSinceOnsetOfSymptomsVector,
+        now: DateTime = Instant().toDateTime(DateTimeZone.UTC)
+    ): List<KeyExportFormat.TemporaryExposureKey> {
+        val result = mutableListOf<KeyExportFormat.TemporaryExposureKey>()
+        keys.groupBy { it.ageInDays(now) }.forEach { entry ->
+            val ageInDays = entry.key
+            entry.value.forEach {
+                result.add(
+                    keyConverter.toExternalFormat(
+                        it,
+                        transmissionRiskVector[ageInDays],
+                        daysSinceOnsetOfSymptomsVector[ageInDays]
+                    )
+                )
+            }
         }
+        return result.toList()
+    }
 
-    fun toSortedHistory(keys: List<TemporaryExposureKey>) =
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun sortWithRecentKeyFirst(keys: List<TemporaryExposureKey>) =
         keys.sortedWith(compareByDescending { it.rollingStartIntervalNumber })
 
-    companion object {
-
-        private const val MAXIMUM_KEYS = 14
-    }
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun TemporaryExposureKey.ageInDays(now: DateTime = Instant().toDateTime(DateTimeZone.UTC)): Int =
```

Comment:
Instant has no time zone. Instant.now() is UTC. As long as the Instant is not from some weird source, we could go without the extra conversion.